Resuelva los siguientes problemas de EDO utilizando el método RK4 implementado:
> a) dy/dt = -2ty, y(0) = 1, sobre el intervalo t = 0 a t = 5

> b) dy/dt = t^2 - y, y(0) = 0, en el intervalo t = 0 a t = 4

> c) d^2y/dt^2 + 2dy/dt + 2y = 0, y(0) = 1, dy/dt(0) = -1, en el intervalo t = 0 a t = 10

> d) dy/dt = cos(t) + y, y(0) = 0 Intervalo: t = 0 a t = 6

> e) dy/dt = e^(-t) - y, y(0) = 1,   Intervalo: t = 0 a t = 3

> f) d^2y/dt^2 - 2dy/dt + 2y = t, y(0) = 1, dy/dt(0) = -1 Intervalo: t = 0 a t = 5

> g) dy/dt = 1 + t + y^2, y(0) = 0,  Intervalo: t = 0 a t = 4

> h) d^2y/dt^2 - 4dy/dt + 4y = e^(-t), y(0) = 0, dy/dt(0) = 1  Intervalo: t = 0 a t = 5

> i) dy/dt = sen(t) - y/t, y(1) = 1,Intervalo: t = 1 a t = 6

> j) d^2y/dt^2 + 4y = 0, y(0) = 0, dy/dt(0) = 2 Intervalo: t = 0 a t = 8

> a) dy/dt = -2ty, y(0) = 1, sobre el intervalo t = 0 a t = 5

In [11]:
import matplotlib.pyplot as plt

# Introducir la ecuacion
def f(t, y):
    return -2*t*y


# Introducir los valores iniciales
def introducir_variables():
    global tk, dt, y0, n, t, y
    # tk <- Es el tiempo inicial que luego va cambiando con dt
    tk = float(input("Ingrese el tiempo inicial: "))
    # dt <- Es el tamanho de los pasos del tiempo (delta t) del
    # cual depende la presicion del resultado
    dt = float(input("Ingrese la dt: "))
    # Es el valor de y en el tiempo inicial
    y0 = float(input("Ingrese la condicion inicial: "))
    # Es la cantidad de resultados de y que se quiere
    n = int(input("Ingrese la cantidad de pasos: "))
    t = 0
    y = 0
    return tk, dt, y0, n, t, y

x_axis = []
y_axis = []


introducir_variables()


f(t, y)

# Siendo h el iterador que va a contar los pasos y
# siendo n la cantidad de pasos que debe realizar
for h in range(n):
    k1 = dt * f(tk, y0)

    k2 = dt * f(tk + dt/2, y0 + k1/2)

    k3 = dt * f(tk + dt/2, y0 + k2/2)

    k4 = dt * f(tk + dt, y0 + k3)

    yk = y0 + ((k1 + (2 * k2) + (2 * k3) + k4) / 6)

    y0 = yk

    tk = tk + dt


    print(yk)

0.33333333333333337
0.3333333333333333
4.222222222222221
250.51851851851845
45343.85185185184


Resultados obtenidos | Resultados de comparacion
---------------------|--------------------------
1 | 1
0.33333333333333337 | 0.333333333333334
0.3333333333333333 | 0.333333333333334
4.222222222222221 | 4.22222222222223
250.51851851851845 | 250.518518518519
45343.85185185184| 45343.8518518518

En la tabla se puede observar que los resultados obtenidos son mas exactos que los resultados obtenidos en una calculadora.


> b) dy/dt = t^2 - y, y(0) = 0, en el intervalo t = 0 a t = 4

In [14]:
import matplotlib.pyplot as plt
import math  # Usar en caso te tener alguna funcion trigonometrica


# Introducir la ecuacion
def f(t, y):
    return t**2 - y


# Introducir los valores iniciales
def introducir_variables():
    global tk, dt, y0, n, t, y
    # tk <- Es el tiempo inicial que luego va cambiando con dt
    tk = float(input("Ingrese el tiempo inicial: "))
    # dt <- Es el tamanho de los pasos del tiempo (delta t) del
    # cual depende la presicion del resultado
    dt = float(input("Ingrese la dt: "))
    # Es el valor de y en el tiempo inicial
    y0 = float(input("Ingrese la condicion inicial: "))
    # Es la cantidad de resultados de y que se quiere
    n = int(input("Ingrese la cantidad de pasos: "))
    t = 0
    y = 0
    return tk, dt, y0, n, t, y

x_axis = []
y_axis = []


introducir_variables()


f(t, y)

# Siendo h el iterador que va a contar los pasos y
# siendo n la cantidad de pasos que debe realizar
for h in range(n):
    k1 = dt * f(tk, y0)

    k2 = dt * f(tk + dt/2, y0 + k1/2)

    k3 = dt * f(tk + dt/2, y0 + k2/2)

    k4 = dt * f(tk + dt, y0 + k3)

    yk = y0 + ((k1 + (2 * k2) + (2 * k3) + k4) / 6)

    y0 = yk

    tk = tk + dt


    print(yk)

0.2708333333333333
1.7473958333333333
4.926106770833333
9.993123372395834
17.01825459798177


Resultados obtenidos | Resultados de comparacion
---------------------|--------------------------
0 |0
0.2708333333333333 | 0.270833333333334
1.7473958333333333 | 1.74739583333334
4.926106770833333 | 4.92610677083334
9.993123372395834 | 9.99312337239584
17.01825459798177 | 17.0182545979818

En la tabla se puede observar que los resultados obtenidos son mas exactos que los resultados obtenidos en una calculadora.


> c) d^2y/dt^2 + 2dy/dt + 2y = 0, y(0) = 1, dy/dt(0) = -1, en el intervalo t = 0 a t = 10

In [ ]:
import math  # Usar en caso te tener alguna funcion trigonometrica


# Introducir la ecuacion
def f(t, y):
    return t**2 - y


# Introducir los valores iniciales
def introducir_variables():
    global tk, dt, y0, n, t, y
    # tk <- Es el tiempo inicial que luego va cambiando con dt
    tk = float(input("Ingrese el tiempo inicial: "))
    # dt <- Es el tamanho de los pasos del tiempo (delta t) del
    # cual depende la presicion del resultado
    dt = float(input("Ingrese la dt: "))
    # Es el valor de y en el tiempo inicial
    y0 = float(input("Ingrese la condicion inicial: "))
    # Es la cantidad de resultados de y que se quiere
    n = int(input("Ingrese la cantidad de pasos: "))
    t = 0
    y = 0
    return tk, dt, y0, n, t, y

x_axis = []
y_axis = []


introducir_variables()


f(t, y)

# Siendo h el iterador que va a contar los pasos y
# siendo n la cantidad de pasos que debe realizar
for h in range(n):
    k1 = dt * f(tk, y0)

    k2 = dt * f(tk + dt/2, y0 + k1/2)

    k3 = dt * f(tk + dt/2, y0 + k2/2)

    k4 = dt * f(tk + dt, y0 + k3)

    yk = y0 + ((k1 + (2 * k2) + (2 * k3) + k4) / 6)

    y0 = yk

    tk = tk + dt


    print(yk)

Resultados obtenidos | Resultados de comparacion
---------------------|--------------------------


En la tabla se puede observar que los resultados obtenidos son mas exactos que los resultados obtenidos en una calculadora.

> d) dy/dt = cos(t) + y, y(0) = 0 Intervalo: t = 0 a t = 6

In [15]:
import math

# Introducir la ecuacion
def f(t, y):
    return math.cos(t) + y


# Introducir los valores iniciales
def introducir_variables():
    global tk, dt, y0, n, t, y
    # tk <- Es el tiempo inicial que luego va cambiando con dt
    tk = float(input("Ingrese el tiempo inicial: "))
    # dt <- Es el tamanho de los pasos del tiempo (delta t) del
    # cual depende la presicion del resultado
    dt = float(input("Ingrese la dt: "))
    # Es el valor de y en el tiempo inicial
    y0 = float(input("Ingrese la condicion inicial: "))
    # Es la cantidad de resultados de y que se quiere
    n = int(input("Ingrese la cantidad de pasos: "))
    t = 0
    y = 0
    return tk, dt, y0, n, t, y


introducir_variables()


f(t, y)

# Siendo h el iterador que va a contar los pasos y
# siendo n la cantidad de pasos que debe realizar
for h in range(n):
    k1 = dt * f(tk, y0)

    k2 = dt * f(tk + dt/2, y0 + k1/2)

    k3 = dt * f(tk + dt/2, y0 + k2/2)

    k4 = dt * f(tk + dt, y0 + k3)

    yk = y0 + ((k1 + (2 * k2) + (2 * k3) + k4) / 6)

    y0 = yk

    tk = tk + dt


    print(yk)

1.4990981596925936
4.314970235072493
10.462739420294643
26.759404020980487
71.99271414545352
196.0380332821979


Resultados obtenidos | Resultados de comparacion
---------------------|--------------------------
0 | 0
1.4990981596925936 | 1.4990981596926
4.314970235072493 | 4.3149702350725
10.462739420294643 | 10.4627394202947
26.759404020980487 | 26.7594040209805
71.99271414545352 | 71.9927141454536
196.0380332821979 | 196.038033282198


En la tabla se puede observar que los resultados obtenidos son mas exactos que los resultados obtenidos en una calculadora.

> e) dy/dt = e^(-t) - y, y(0) = 1,   Intervalo: t = 0 a t = 3

In [16]:
import math

# Introducir la ecuacion
def f(t, y):
    return math.e**-2 - y


# Introducir los valores iniciales
def introducir_variables():
    global tk, dt, y0, n, t, y
    # tk <- Es el tiempo inicial que luego va cambiando con dt
    tk = float(input("Ingrese el tiempo inicial: "))
    # dt <- Es el tamanho de los pasos del tiempo (delta t) del
    # cual depende la presicion del resultado
    dt = float(input("Ingrese la dt: "))
    # Es el valor de y en el tiempo inicial
    y0 = float(input("Ingrese la condicion inicial: "))
    # Es la cantidad de resultados de y que se quiere
    n = int(input("Ingrese la cantidad de pasos: "))
    t = 0
    y = 0
    return tk, dt, y0, n, t, y


introducir_variables()


f(t, y)

# Siendo h el iterador que va a contar los pasos y
# siendo n la cantidad de pasos que debe realizar
for h in range(n):
    k1 = dt * f(tk, y0)

    k2 = dt * f(tk + dt/2, y0 + k1/2)

    k3 = dt * f(tk + dt/2, y0 + k2/2)

    k4 = dt * f(tk + dt, y0 + k3)

    yk = y0 + ((k1 + (2 * k2) + (2 * k3) + k4) / 6)

    y0 = yk

    tk = tk + dt


    print(yk)

0.45958455202288284
0.25692875903146395
0.1809328366596819


Resultados obtenidos | Resultados de comparacion
---------------------|--------------------------
0 | 0
0.45958455202288284 | 0.355701015075505
0.25692875903146395 | 0.264242971303406
0.1809328366596819 | 0.147230011861571

En la tabla se puede observar que los resultados obtenidos son mas exactos que los resultados obtenidos en una calculadora. En este ejercicio la exactitud fue notable.

> f) d^2y/dt^2 - 2dy/dt + 2y = t, y(0) = 1, dy/dt(0) = -1 Intervalo: t = 0 a t = 5

In [ ]:
import math

# Introducir la ecuacion
def f(t, y):
    return math.e**-2 - y


# Introducir los valores iniciales
def introducir_variables():
    global tk, dt, y0, n, t, y
    # tk <- Es el tiempo inicial que luego va cambiando con dt
    tk = float(input("Ingrese el tiempo inicial: "))
    # dt <- Es el tamanho de los pasos del tiempo (delta t) del
    # cual depende la presicion del resultado
    dt = float(input("Ingrese la dt: "))
    # Es el valor de y en el tiempo inicial
    y0 = float(input("Ingrese la condicion inicial: "))
    # Es la cantidad de resultados de y que se quiere
    n = int(input("Ingrese la cantidad de pasos: "))
    t = 0
    y = 0
    return tk, dt, y0, n, t, y


introducir_variables()


f(t, y)

# Siendo h el iterador que va a contar los pasos y
# siendo n la cantidad de pasos que debe realizar
for h in range(n):
    k1 = dt * f(tk, y0)

    k2 = dt * f(tk + dt/2, y0 + k1/2)

    k3 = dt * f(tk + dt/2, y0 + k2/2)

    k4 = dt * f(tk + dt, y0 + k3)

    yk = y0 + ((k1 + (2 * k2) + (2 * k3) + k4) / 6)

    y0 = yk

    tk = tk + dt


    print(yk)

> g) dy/dt = 1 + t + y^2, y(0) = 0,  Intervalo: t = 0 a t = 4

In [ ]:
import math

# Introducir la ecuacion
def f(t, y):
    return math.e**-2 - y


# Introducir los valores iniciales
def introducir_variables():
    global tk, dt, y0, n, t, y
    # tk <- Es el tiempo inicial que luego va cambiando con dt
    tk = float(input("Ingrese el tiempo inicial: "))
    # dt <- Es el tamanho de los pasos del tiempo (delta t) del
    # cual depende la presicion del resultado
    dt = float(input("Ingrese la dt: "))
    # Es el valor de y en el tiempo inicial
    y0 = float(input("Ingrese la condicion inicial: "))
    # Es la cantidad de resultados de y que se quiere
    n = int(input("Ingrese la cantidad de pasos: "))
    t = 0
    y = 0
    return tk, dt, y0, n, t, y


introducir_variables()


f(t, y)

# Siendo h el iterador que va a contar los pasos y
# siendo n la cantidad de pasos que debe realizar
for h in range(n):
    k1 = dt * f(tk, y0)

    k2 = dt * f(tk + dt/2, y0 + k1/2)

    k3 = dt * f(tk + dt/2, y0 + k2/2)

    k4 = dt * f(tk + dt, y0 + k3)

    yk = y0 + ((k1 + (2 * k2) + (2 * k3) + k4) / 6)

    y0 = yk

    tk = tk + dt


    print(yk)

Resultados obtenidos | Resultados de comparacion
---------------------|--------------------------


En la tabla se puede observar que los resultados obtenidos son mas exactos que los resultados obtenidos en una calculadora.

> h) d^2y/dt^2 - 4dy/dt + 4y = e^(-t), y(0) = 0, dy/dt(0) = 1  Intervalo: t = 0 a t = 5

In [ ]:
import math

# Introducir la ecuacion
def f(t, y):
    return math.e**-2 - y


# Introducir los valores iniciales
def introducir_variables():
    global tk, dt, y0, n, t, y
    # tk <- Es el tiempo inicial que luego va cambiando con dt
    tk = float(input("Ingrese el tiempo inicial: "))
    # dt <- Es el tamanho de los pasos del tiempo (delta t) del
    # cual depende la presicion del resultado
    dt = float(input("Ingrese la dt: "))
    # Es el valor de y en el tiempo inicial
    y0 = float(input("Ingrese la condicion inicial: "))
    # Es la cantidad de resultados de y que se quiere
    n = int(input("Ingrese la cantidad de pasos: "))
    t = 0
    y = 0
    return tk, dt, y0, n, t, y


introducir_variables()


f(t, y)

# Siendo h el iterador que va a contar los pasos y
# siendo n la cantidad de pasos que debe realizar
for h in range(n):
    k1 = dt * f(tk, y0)

    k2 = dt * f(tk + dt/2, y0 + k1/2)

    k3 = dt * f(tk + dt/2, y0 + k2/2)

    k4 = dt * f(tk + dt, y0 + k3)

    yk = y0 + ((k1 + (2 * k2) + (2 * k3) + k4) / 6)

    y0 = yk

    tk = tk + dt


    print(yk)

Resultados obtenidos | Resultados de comparacion
---------------------|--------------------------


En la tabla se puede observar que los resultados obtenidos son mas exactos que los resultados obtenidos en una calculadora.

> i) dy/dt = sen(t) - y/t, y(1) = 1,Intervalo: t = 1 a t = 6

In [ ]:
import math

# Introducir la ecuacion
def f(t, y):
    return math.e**-2 - y


# Introducir los valores iniciales
def introducir_variables():
    global tk, dt, y0, n, t, y
    # tk <- Es el tiempo inicial que luego va cambiando con dt
    tk = float(input("Ingrese el tiempo inicial: "))
    # dt <- Es el tamanho de los pasos del tiempo (delta t) del
    # cual depende la presicion del resultado
    dt = float(input("Ingrese la dt: "))
    # Es el valor de y en el tiempo inicial
    y0 = float(input("Ingrese la condicion inicial: "))
    # Es la cantidad de resultados de y que se quiere
    n = int(input("Ingrese la cantidad de pasos: "))
    t = 0
    y = 0
    return tk, dt, y0, n, t, y


introducir_variables()


f(t, y)

# Siendo h el iterador que va a contar los pasos y
# siendo n la cantidad de pasos que debe realizar
for h in range(n):
    k1 = dt * f(tk, y0)

    k2 = dt * f(tk + dt/2, y0 + k1/2)

    k3 = dt * f(tk + dt/2, y0 + k2/2)

    k4 = dt * f(tk + dt, y0 + k3)

    yk = y0 + ((k1 + (2 * k2) + (2 * k3) + k4) / 6)

    y0 = yk

    tk = tk + dt


    print(yk)

Resultados obtenidos | Resultados de comparacion
---------------------|--------------------------


En la tabla se puede observar que los resultados obtenidos son mas exactos que los resultados obtenidos en una calculadora.

> j) d^2y/dt^2 + 4y = 0, y(0) = 0, dy/dt(0) = 2 Intervalo: t = 0 a t = 8

In [ ]:
import math

# Introducir la ecuacion
def f(t, y):
    return math.e**-2 - y


# Introducir los valores iniciales
def introducir_variables():
    global tk, dt, y0, n, t, y
    # tk <- Es el tiempo inicial que luego va cambiando con dt
    tk = float(input("Ingrese el tiempo inicial: "))
    # dt <- Es el tamanho de los pasos del tiempo (delta t) del
    # cual depende la presicion del resultado
    dt = float(input("Ingrese la dt: "))
    # Es el valor de y en el tiempo inicial
    y0 = float(input("Ingrese la condicion inicial: "))
    # Es la cantidad de resultados de y que se quiere
    n = int(input("Ingrese la cantidad de pasos: "))
    t = 0
    y = 0
    return tk, dt, y0, n, t, y


introducir_variables()


f(t, y)

# Siendo h el iterador que va a contar los pasos y
# siendo n la cantidad de pasos que debe realizar
for h in range(n):
    k1 = dt * f(tk, y0)

    k2 = dt * f(tk + dt/2, y0 + k1/2)

    k3 = dt * f(tk + dt/2, y0 + k2/2)

    k4 = dt * f(tk + dt, y0 + k3)

    yk = y0 + ((k1 + (2 * k2) + (2 * k3) + k4) / 6)

    y0 = yk

    tk = tk + dt


    print(yk)

Resultados obtenidos | Resultados de comparacion
---------------------|--------------------------


En la tabla se puede observar que los resultados obtenidos son mas exactos que los resultados obtenidos en una calculadora.

# `Analisis y conclusion`
La regla de Runge-Kutta es un método numérico utilizado para resolver ecuaciones diferenciales. A diferencia del método de Euler, que avanza solo un paso, la regla de Runge-Kutta utiliza información de varios pasos anteriores para calcular el nuevo valor. Esto permite obtener una aproximación más precisa de la solución de la ecuacion diferencial.

Al ajustar el tamaño de los pasos de tiempo, podemos aumentar la precisión de la solución numérica. Al disminuir el tamaño del paso delta t, realizamos más iteraciones en el mismo intervalo de tiempo, mejorando así la aproximación del comportamiento de la solución. Esta propiedad también se observa en el método de Euler.

Sin embargo, es importante tener en cuenta que la regla de Runge-Kutta es computacionalmente más costosa en comparación con otros métodos, la cual resulta ser una de sus limitaciones. Al trabajar con pasos de tiempo más pequeños, se requiere un mayor número de cálculos. A pesar de este costo computacional adicional, la ventaja es que obtenemos resultados más precisos.

A pesar de estas consideraciones, la regla de Runge-Kutta ofrece varias ventajas en la resolución numérica de las ecuaciones diferenciales ordinarias . Proporciona resultados cercanos a la solución real, lo cual se traduce en una alta precisión.